jobserver/views/users.py

```diff
@@ -1,82 +1,64 @@
 
     def form_valid(self, form):
         user = User.objects.filter(email=form.cleaned_data["email"]).first()
-
-        # we have three possible states for the given email:
-        # * unknown user
-        # * GitHub user
-        # * email-only user
-        # we want to show the same outcome regardless of state so a bad actor
-        # can't work out work out who uses the service by email
         if user:
-            if user.social_auth.exists():
-                # send GitHub users a link to the PSA entrypoint and explain
-                # that they don't have a password to reset
-                send_github_login_email(user)
-            else:
-                send_reset_password_email(user)
+            send_login_email(user)
 
-        messages.success(
-            self.request, "Your password reset request was successfully sent."
-        )
-        return redirect("/")
+        messages.success(self.request, "Check your email for a login link")
 
+        context = self.get_context_data(next_url=self.next_url)
+        return self.render_to_response(context)
 
-class SetPassword(PasswordResetConfirmView):
-    post_reset_login = True
-    post_reset_login_backend = "django.contrib.auth.backends.ModelBackend"
-    success_url = "/"
-    template_name = "set_password.html"
+    def get_context_data(self, **kwargs):
+        return super().get_context_data(**kwargs) | {
+            "next_url": self.next_url,
+        }
 
-    def get_success_url(self):
-        # the authorization framework doesn't support getting all roles for a
-        # user across all objects, so we hae to manually do it here.
-        roles = set(self.request.user.roles)
-        for membership in self.request.user.project_memberships.all():
-            roles |= set(membership.roles)
-
-        # treat users with only the InteractiveReporter role as interactive users.
-        # we'll have users with this role along with others, eg for staff, but
-        # they likely don't want to be redirected on login.
-        if roles != {InteractiveReporter}:
-            return "/"
-
-        # if the user is logging in via this view, so with an email address and
-        # password, _and_ has the InteractiveReporter role then they should
-        # definitely be an interactive user.
-
-        # TODO: add a page for users who have 2+ projects
-        project = self.request.user.projects.first()
-        return project.get_interactive_url()
+    def get_next_url(self, request):
+        next_url = request.GET.get("next") or "/"
+        if not is_safe_path(next_url):
+            next_url = ""
+
+        return next_url
+
+
+class LoginWithURL(View):
+    def get(self, request, *args, **kwargs):
+        try:
+            user = User.from_login_token(self.kwargs["token"])
+        except (User.DoesNotExist) as e:
+            print(e)
+            messages.error(request, "An error occurred, please try again")
```

Comment:
```suggestion
            messages.error(request, "Invalid token, please try again")
```

Reply:
Should we expose that there was a problem with the token specifically?  I'm not sure it adds anything meaningful for the user to know this, it's not like they can do anything about it!

Reply:
I think it does.

"Error" suggests something went wrong on our end, possibly nothing to do with their login attempt.

But this is either a token mismatch, or an expired token.  The former is also what happens when they click on the same url again.

"An error" is just too open to be useful to guide them what went wrong